File: skills-discovery/backend/app/routers/collections.py
from fastapi import APIRouter, Depends, HTTPException
from typing import List
import aiosqlite

from ..database import get_db
from ..models import CollectionCreate, CollectionOut, SkillOut, TagOut

router = APIRouter()


async def fetch_tags_for_skill(db: aiosqlite.Connection, skill_id: int) -> List[TagOut]:
    async with db.execute(
        """SELECT t.id, t.name FROM tags t
           JOIN skill_tags st ON st.tag_id = t.id
           WHERE st.skill_id = ?""",
        (skill_id,)
    ) as cursor:
        rows = await cursor.fetchall()
        return [TagOut(id=r["id"], name=r["name"]) for r in rows]


async def row_to_skill(db: aiosqlite.Connection, row) -> SkillOut:
    tags = await fetch_tags_for_skill(db, row["id"])
    async with db.execute("SELECT 1 FROM favorites WHERE skill_id = ?", (row["id"],)) as fav:
        is_fav = await fav.fetchone() is not None
    return SkillOut(
        id=row["id"],
        name=row["name"],
        description=row["description"],
        category=row["category"],
        source_url=row["source_url"],
        source_name=row["source_name"],
        pricing=row["pricing"] or "free",
        price_details=row["price_details"],
        features=row["features"] or "[]",
        install_instructions=row["install_instructions"],
        version=row["version"],
        last_checked=row["last_checked"],
        is_active=row["is_active"],
        popularity_score=row["popularity_score"] or 0.0,
        created_at=row["created_at"],
        updated_at=row["updated_at"],
        tags=tags,
        is_favorite=is_fav,
    )


async def build_collection_out(db: aiosqlite.Connection, row) -> CollectionOut:
    async with db.execute(
        """SELECT s.* FROM skills s
           JOIN collection_skills cs ON cs.skill_id = s.id
           WHERE cs.collection_id = ? AND s.is_active = 1""",
        (row["id"],)
    ) as cursor:
        skill_rows = await cursor.fetchall()
    skills = [await row_to_skill(db, sr) for sr in skill_rows]
    return CollectionOut(
        id=row["id"],
        name=row["name"],
        description=row["description"],
        created_at=row["created_at"],
        skills=skills,
    )


@router.get("", response_model=List[CollectionOut])
async def list_collections(db: aiosqlite.Connection = Depends(get_db)):
    async with db.execute("SELECT * FROM collections ORDER BY created_at DESC") as cursor:
        rows = await cursor.fetchall()
    return [await build_collection_out(db, row) for row in rows]


@router.get("/{collection_id}", response_model=CollectionOut)
async def get_collection(collection_id: int, db: aiosqlite.Connection = Depends(get_db)):
    async with db.execute("SELECT * FROM collections WHERE id = ?", (collection_id,)) as cursor:
        row = await cursor.fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Collection not found")
    return await build_collection_out(db, row)


@router.post("", response_model=CollectionOut, status_code=201)
async def create_collection(coll_in: CollectionCreate, db: aiosqlite.Connection = Depends(get_db)):
    cursor = await db.execute(
        "INSERT INTO collections (name, description) VALUES (?, ?)",
        (coll_in.name, coll_in.description)
    )
    coll_id = cursor.lastrowid
    await db.commit()
    async with db.execute("SELECT * FROM collections WHERE id = ?", (coll_id,)) as cur:
        row = await cur.fetchone()
    return await build_collection_out(db, row)


@router.put("/{collection_id}", response_model=CollectionOut)
async def update_collection(
    collection_id: int,
    coll_in: CollectionCreate,
    db: aiosqlite.Connection = Depends(get_db),
):
    async with db.execute("SELECT id FROM collections WHERE id = ?", (collection_id,)) as cursor:
        row = await cursor.fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Collection not found")
    await db.execute(
        "UPDATE collections SET name = ?, description = ? WHERE id = ?",
        (coll_in.name, coll_in.description, collection_id)
    )
    await db.commit()
    async with db.execute("SELECT * FROM collections WHERE id = ?", (collection_id,)) as cur:
        row = await cur.fetchone()
    return await build_collection_out(db, row)


@router.delete("/{collection_id}", status_code=204)
async def delete_collection(collection_id: int, db: aiosqlite.Connection = Depends(get_db)):
    async with db.execute("SELECT id FROM collections WHERE id = ?", (collection_id,)) as cursor:
        row = await cursor.fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Collection not found")
    await db.execute("DELETE FROM collections WHERE id = ?", (collection_id,))
    await db.commit()


@router.post("/{collection_id}/skills/{skill_id}", status_code=201)
async def add_skill_to_collection(
    collection_id: int, skill_id: int, db: aiosqlite.Connection = Depends(get_db)
):
    async with db.execute("SELECT id FROM collections WHERE id = ?", (collection_id,)) as cursor:
        if not await cursor.fetchone():
            raise HTTPException(status_code=404, detail="Collection not found")
    async with db.execute("SELECT id FROM skills WHERE id = ?", (skill_id,)) as cursor:
        if not await cursor.fetchone():
            raise HTTPException(status_code=404, detail="Skill not found")
    await db.execute(
        "INSERT OR IGNORE INTO collection_skills (collection_id, skill_id) VALUES (?, ?)",
        (collection_id, skill_id)
    )
    await db.commit()
    return {"added": True}


@router.delete("/{collection_id}/skills/{skill_id}", status_code=204)
async def remove_skill_from_collection(
    collection_id: int, skill_id: int, db: aiosqlite.Connection = Depends(get_db)
):
    await db.execute(
        "DELETE FROM collection_skills WHERE collection_id = ? AND skill_id = ?",
        (collection_id, skill_id)
    )
    await db.commit()


@router.get("/favorites/list", response_model=List[SkillOut])
async def get_favorites(db: aiosqlite.Connection = Depends(get_db)):
    async with db.execute(
        """SELECT s.* FROM skills s
           JOIN favorites f ON f.skill_id = s.id
           WHERE s.is_active = 1
           ORDER BY f.created_at DESC"""
    ) as cursor:
        rows = await cursor.fetchall()
    return [await row_to_skill(db, row) for row in rows]

File: skills-discovery/api/src/routes/collections.js
import { Hono } from 'hono'
import { db } from '../db.js'

const router = new Hono()

// Skills favoris — déclaré avant /:id pour éviter la collision de route
router.get('/favorites/list', (c) => {
  const skills = db
    .prepare(
      `SELECT s.* FROM skills s JOIN favorites f ON f.skill_id = s.id
       ORDER BY f.created_at DESC`
    )
    .all()
  const getTags = db.prepare(
    'SELECT t.name FROM tags t JOIN skill_tags st ON st.tag_id = t.id WHERE st.skill_id = ?'
  )
  for (const s of skills) {
    s.tags = getTags.all(s.id).map((r) => r.name)
    s.is_favorite = 1
  }
  return c.json(skills)
})

router.get('/', (c) => {
  const collections = db
    .prepare(
      `SELECT col.*, COUNT(cs.skill_id) AS skill_count
       FROM collections col
       LEFT JOIN collection_skills cs ON cs.collection_id = col.id
       GROUP BY col.id ORDER BY col.created_at DESC`
    )
    .all()
  return c.json(collections)
})

router.get('/:id', (c) => {
  const id = Number(c.req.param('id'))
  const collection = db.prepare('SELECT * FROM collections WHERE id = ?').get(id)
  if (!collection) return c.json({ error: 'Collection introuvable' }, 404)
  collection.skills = db
    .prepare(
      `SELECT s.* FROM skills s JOIN collection_skills cs ON cs.skill_id = s.id
       WHERE cs.collection_id = ? ORDER BY s.name`
    )
    .all(id)
  return c.json(collection)
})

router.post('/', async (c) => {
  const { name, description } = await c.req.json()
  if (!name?.trim()) return c.json({ error: 'name requis' }, 400)
  const info = db
    .prepare('INSERT INTO collections (name, description) VALUES (?, ?)')
    .run(name.trim(), description || null)
  return c.json(db.prepare('SELECT * FROM collections WHERE id = ?').get(info.lastInsertRowid), 201)
})

router.put('/:id', async (c) => {
  const id = Number(c.req.param('id'))
  const col = db.prepare('SELECT * FROM collections WHERE id = ?').get(id)
  if (!col) return c.json({ error: 'Collection introuvable' }, 404)
  const body = await c.req.json()
  db.prepare('UPDATE collections SET name = ?, description = ? WHERE id = ?').run(
    body.name ?? col.name,
    body.description ?? col.description,
    id
  )
  return c.json(db.prepare('SELECT * FROM collections WHERE id = ?').get(id))
})

router.delete('/:id', (c) => {
  const info = db.prepare('DELETE FROM collections WHERE id = ?').run(Number(c.req.param('id')))
  if (!info.changes) return c.json({ error: 'Collection introuvable' }, 404)
  return c.body(null, 204)
})

router.post('/:id/skills/:skillId', (c) => {
  const id = Number(c.req.param('id'))
  const skillId = Number(c.req.param('skillId'))
  if (!db.prepare('SELECT 1 FROM collections WHERE id = ?').get(id))
    return c.json({ error: 'Collection introuvable' }, 404)
  if (!db.prepare('SELECT 1 FROM skills WHERE id = ?').get(skillId))
    return c.json({ error: 'Skill introuvable' }, 404)
  db.prepare('INSERT OR IGNORE INTO collection_skills (collection_id, skill_id) VALUES (?, ?)').run(id, skillId)
  return c.json({ collection_id: id, skill_id: skillId }, 201)
})

router.delete('/:id/skills/:skillId', (c) => {
  const info = db
    .prepare('DELETE FROM collection_skills WHERE collection_id = ? AND skill_id = ?')
    .run(Number(c.req.param('id')), Number(c.req.param('skillId')))
  if (!info.changes) return c.json({ error: 'Association introuvable' }, 404)
  return c.body(null, 204)
})

export default router
